### PR TITLE
Re-updating UI

### PR DESCRIPTION
**Pull request by ryanbugden**
I made UI updates similar to these in Sept 2021, but I did them off of Stephen's branch, so I think they never saw the light of day. Hopefully this works; Github is being confusing.

I tried to make the window more compact, sleeker, and a bit more organized spatially, while also coding it such that if a new checkbox is added, for instance, the window height will automatically increase.

Please see commit descriptions + comments on the code for more context.

Thanks!
Ryan

### CODE REVIEW

File: MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py [hunk 1]
```diff
@@ -15,139 +15,147 @@
 from defconAppKit.windows.baseWindow import BaseWindowController
 from mojo.events import addObserver, removeObserver, postEvent
 from vanilla import *
+from lib.UI.integerEditText import NumberEditText
```
Comment (ryanbugden) [inline MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py:18]: NumberEditText is what RF uses in the Glyph Info pane in Inspector.

File: MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py [hunk 2]
```diff
@@ -15,139 +15,147 @@
...
 import codecs
 
 class MM2SpaceCenter:
     
     def activateModule(self):
         addObserver(self, "MMPairChangedObserver", "MetricsMachine.currentPairChanged")
+        
         print ('MM2SpaceCenter activated')
 
+
```
Comment (ryanbugden) [inline MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py:29]: I added/removed some line breaks for a bit of consistency and to tighten the gaps so I could see indent relationships better. Hope that's ok!

File: MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py [hunk 3]
```diff
@@ -15,139 +15,147 @@
...
     def deactivateModule(self, sender):
         removeObserver(self, "MetricsMachine.currentPairChanged")
+        removeObserver(self, "FutureProof.currentPairChanged")
```
Comment (ryanbugden) [inline MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py:32]: I don't know what this is. Github was being unintuitive, so I enacted my UI updates upon the latest version of the file I could find.

File: MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py [hunk 4]
```diff
@@ -15,139 +15,147 @@
...
+        
+        leftMargin = 11
+        topMargin = 0
+        lineHeight = 18
+        columnWidth = 90  
+        second_col_x = columnWidth + leftMargin + 10
 
+        w_w = 200
         
+        self.messageText = '😎 Activated'
```
Comment (ryanbugden) [inline MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py:55]: Title bar already said the name of the extension, so I figured some brevity here would be nice.

File: MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py [hunk 5]
```diff
@@ -15,139 +15,147 @@
...
+
         self.wordCount = 20
         self.minLength = 3
         self.maxLength = 15
         
         self.activateModule()
-        self.w = Window((250, 155), "MM2SpaceCenter")
+        self.w = Window((w_w, 0), "MM2SpaceCenter")
```
Comment (ryanbugden) [inline MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py:62]: Height starts at `0`, and it's determined after all UI elements are made.

File: MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py [hunk 6]
```diff
@@ -15,139 +15,147 @@
...
+        self.w.getNSWindow().setTitlebarHeight_(26)
+        self.w.getNSWindow().setTitlebarAppearsTransparent_(True)
```
Comment (ryanbugden) [inline MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py:64]: Smaller and sleeker title bar.

File: MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py [hunk 7]
```diff
@@ -15,139 +15,147 @@
...
         
-        self.w.myTextBox = TextBox((leftMargin, yPos, -10, 17), self.messageText, sizeStyle="regular") 
-
+        yPos = topMargin
 
+        self.w.statusBar = Box((3, 0, -3, 30))
```
Comment (ryanbugden) [inline MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py:68]: Status bar feels like a status hierarchically. Helps to delineate areas in this tight space.

File: MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py [hunk 8]
```diff
@@ -15,139 +15,147 @@
...
         self.source = None 
         self.source = self.w.source.get() #get value, to use for other functions
-
-
-        yPos += lineHeight * 1.2
+        
+        ##### add drop down to choose context: auto, UC, lc, figs, ??, not functional yet, but made as a placeholder
+        
+        self.contextOptions = ['Auto', 'UC', 'LC', 'Figs', 'Frac']
+        self.w.context = PopUpButton((second_col_x, yPos, -leftMargin, 20), [], sizeStyle="small", callback=self.changeContextCallback)
```
Comment (ryanbugden) [inline MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py:104]: Second column x position is defined, and the second drop down menu automatically goes to right margin.

File: MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py [hunk 9]
```diff
@@ -346,27 +360,83 @@ def getPairstring(self, pair):
...
         try:
             #print ('pair =', pair)
-            left = self.gname2char(CurrentFont(), pair[0])
-            right = self.gname2char(CurrentFont(), pair[1])
+
+            leftNoSuffix = pair[0]
+            rightNoSuffix = pair[1]
+
+            leftPeriodPos = pair[0].find(".")
+            if leftPeriodPos > 0:
```
Comment (ryanbugden) [inline MM2SpaceCenter.roboFontExt/lib/MM2SpaceCenter.py:375]: Don't know what this is. Inherited from other file.

**Comment by cjdunn** [comment]
Thanks @ryanbugden for looking into the UI. I think your naming and organization of the code is probably better than what I had in there now. But the overall result I don't think is an improvement. I previously tried making the current UI narrower as you have it, but I left it longer so it would show glyph names for unencoded glyphs like /hyphen.uc etc (see screenshot)
show_unencoded_glyph_names

Ideally this text could be truncated if it's too long, which isn't happening in the current version or in your version, but I think would be best.**

Vertically, yours ends up being longer, which is not preferable for me. I don't think the horizontal bar helps much and I don't think the darker grey box behind the messageText is necessary. 

**After writing these comments I made a version which implements your naming/structural updates, but keeps the wider width that I previously had and shorter length. I also decided to finally tackle the truncation issue and I think it's working pretty well now (see screenshot) 
truncation

Thanks for the work you did on this, and the motivation to look closer at the UI! 

My updated branch, which I will merge into next-master is here: 
https://github.com/cjdunn/MM2SpaceCenter/tree/fea-UI-update

**Comment by ryanbugden** [comment]
Hey CJ, thanks for the consideration. 

I think little things like the status bar help to create a separation of information (not confusing the status with the title bar or text immediately below). The eye may go back and forth periodically between the space center and this UI, so I think a little extra height is worth it for a more legible organization of information. The status bar now steals some vertical real estate from the title bar, which can afford it because it's now transparent and less constrained. Another reason for a status bar is that it's the only piece of information that is dynamic and non-editable (status). Having a different visual treatment for information that's innately different I think is helpful telegraphically.

Other things were niceties, like allowing the # EditText box to be wider, in favor of "words" moving to the right and justifying with the first drop-down menu.

As for the horizontal line, this is optional, but I would still recommend a bit more space there, to help compete with the space  between the two drop down menus, for instance.

Re: keeping the width of the status string compact: you may consider removing the emojis altogether, or only placing an emoji (sad/red/etc) as a shorthand for "not found" and then add the pair.

All said, I understand that there's an element of subjectivity here, so feel free to ignore. Glad you found some of the structural changes (NumberEditText, resizing height) useful.

**Pull request closed without merging**